Improve typing coverage on external API, in particular Worker and Broker classes

jenstroeger opened the pull request:
Closes #727

Adds a bunch of type annotations to many public functions; in two cases the annotations would require a minor rewrite of code itself.

I’ve not annotated `message` arguments because they can be either of type `Message` or `MessageProxy` (because they’re used interchangeably). I also refrained from importing modules guarded by `if TYPE_CHECKING` but I can add that too and further improve some method signatures. Please let me know what you prefer!

synweap15 commented:
I don't see anything wrong with `if typing.TYPE_CHECKING`, feel free to go for it. I'd aim for max typing coverage, so I'd suggest to include `message` in the annotation efforts.

I'm waiting for Python 3.9 EOL to get green light for using `type | None` :)

LincolnPuzey commented on dramatiq/broker.py:
```diff
@@ -317,7 +317,7 @@ class Consumer:
     """
 
-    def __iter__(self):  # pragma: no cover
+    def __iter__(self) -> "Consumer":  # pragma: no cover
```
I think the return type here should be `typing_extensions.Self` because in the subclasses, the return type will be the subclass, not the parent class `Consumer`.

jenstroeger replied:
Unfortunately, [`Self`](https://docs.python.org/3/library/typing.html#typing.Self) was added with 3.11 but this change wants to type-check on 3.9 and up.

LincolnPuzey replied:
I believe if you import it from `typing_extensions` inside of `if TYPE_CHECKING` like [here](https://github.com/Bogdanp/dramatiq/blob/057e03b1727c2813c6e8f5ba209bd3cbec786b0d/dramatiq/actor.py#L41-L42) it should work on 3.9

LincolnPuzey commented on dramatiq/brokers/stub.py:
```diff
@@ -157,8 +159,8 @@ def join(self, queue_name, *, fail_fast=False, timeout=None):
             for queue in queues:
-                timeout = deadline and deadline - time.monotonic()
-                join_queue(queue, timeout=timeout)
+                timeout_ = deadline and deadline - time.monotonic()
```
maybe `timeout_` -> `join_timeout` for clarity?

LincolnPuzey reviewed (commented):
Looks good so far, left some initial comments

jenstroeger commented on dramatiq/results/backend.py:
```diff
@@ -97,11 +97,11 @@ def get_result(self, message, *, block: bool = False, timeout: typing.Optional[i
                 attempts, delay = compute_backoff(attempts, factor=BACKOFF_FACTOR)
-                delay /= 1000
-                if time.monotonic() + delay > end_time:
+                delay_ = delay / 1000
```
Or we’d [allow redefining variables](https://mypy.readthedocs.io/en/stable/config_file.html#confval-allow_redefinition)…

synweap15 replied:
Here we change the unit so a separate var is justified. Redefining variables is discouraged with an occasional `# mypy: ignore` where appropriate & justified

jenstroeger commented on dramatiq/worker.py:
```diff
@@ -332,11 +360,12 @@ def handle_message(self, message):
         individual messages, signaling that each message is ready to
         be acked or rejected.
         """
+        assert self.consumer
```
Guard this with an `if TYPE_CHECKING` to show that this assertion is for the benefit of static type checking only?

jenstroeger replied:
It still understands `asserts` and uses them for [type narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html#type-narrowing-expressions). Looks like 
```python
if TYPE_CHECKING:
    assert self.consumer 
```
works just fine and at least makes the intent of the `assert` clearer to the reader. What do you think?

LincolnPuzey replied:
Ok good point about the type narrowing. This is good then.

jenstroeger commented on dramatiq/broker.py:
```diff
@@ -365,12 +379,24 @@ def close(self) -> None:
+    # For the purpose of static type checking we duplicate the fields
+    # and their respective types of the Message here; at runtime the
+    # message's fields are accessed through the __getattr__ method.
+    if TYPE_CHECKING:
```
Fairnuff. I do prefer the guard to make the intent more explicit: the block of type declarations is for static type checking only.

jenstroeger commented:
I addressed the above comments. Shall we noodle through the middlewares as well, or are they not part of Dramatiq’s external interface? I think I’m easy either way.

LincolnPuzey reviewed (approved):
Thanks @jenstroeger Looks good to me.

LincolnPuzey commented:
>  Shall we noodle through the middlewares as well, or are they not part of Dramatiq’s external interface? I think I’m easy either way.

Lets not add to this PR any more. Another PR for the base middleware class especially would be good though (#521)
